django_irods/views.py

```diff
@@ -0,0 +1,477 @@
+
+    if use_reverse_proxy and getattr(settings, 'SENDFILE_ON', False) and \
+       'HTTP_X_DJANGO_REVERSE_PROXY' in request.META:
+
+        # The NGINX sendfile abstraction is invoked as follows:
+        # 1. The request to download a file enters this routine via the /rest_download or /download
+        #    url in ./urls.py. It is redirected here from Django. The URI contains either the
+        #    unqualified resource path or the federated resource path, depending upon whether
+        #    the request is local or federated.
+        # 2. This deals with unfederated resources by redirecting them to the uri
+        #    /irods-data/{resource-id}/... on nginx. This URI is configured to read the file
+        #    directly from the iRODS vault via NFS, and does not work for direct access to the
+        #    vault due to the 'internal;' declaration in NGINX.
+        # 3. This deals with federated resources by reading their path, matching local vaults, and
+        #    redirecting to URIs that are in turn mapped to read from appropriate iRODS vaults. At
+        #    present, the only one of these is /irods-user, which handles files whose federation
+        #    path is stored in the variable 'userpath'.
+        # 4. If there is no vault available for the resource, the file is transferred without
+        #    NGINX, exactly as it was transferred previously.
+
+        # If this path is resource_federation_path, then the file is a local user file
+        userpath = '/' + os.path.join(
+            getattr(settings, 'HS_USER_IRODS_ZONE', 'hydroshareuserZone'),
+            'home',
+            getattr(settings, 'HS_LOCAL_PROXY_USER_IN_FED_ZONE', 'localHydroProxy'))
+
+        # stop NGINX targets that are non-existent from hanging forever.
+        if not istorage.exists(irods_output_path):
+            content_msg = "file path {} does not exist in iRODS".format(output_path)
+            response = HttpResponse(status=404)
+            if rest_call:
+                response.content = content_msg
+            else:
+                response.content = "<h1>" + content_msg + "</h1>"
+            return response
+
+        if not res.is_federated:
+            # invoke X-Accel-Redirect on physical vault file in nginx
+            response = HttpResponse(content_type=mtype)
+            response['Content-Disposition'] = 'attachment; filename="{name}"'.format(
+                name=output_path.split('/')[-1])
+            response['Content-Length'] = flen
+            response['X-Accel-Redirect'] = '/'.join([
+                getattr(settings, 'IRODS_DATA_URI', '/irods-data'), output_path])
+            logger.debug("Reverse proxying local {}".format(response['X-Accel-Redirect']))
+            return response
+
+        elif res.resource_federation_path == userpath:  # this guarantees a "user" resource
+            # invoke X-Accel-Redirect on physical vault file in nginx
```

Comment:
Although this guarantee a "user" resource, it prevents third-party federated resource from being downloaded. Are you planning to handle external third party federated resource in the similar way as "user" resource? Do we really need to add this constraint here?